cowbird/services/impl/geoserver.py

```diff
@@ -50,42 +59,335 @@ def permission_deleted(self, permission):
-        # TODO
-        return 1
 
-    def create_datastore(self, workspace_id, name):
-        # type (Geoserver, int, str) -> int
+        response = self._create_workspace_request(name)
+        response_code = response.status_code
+        string_to_find = "Workspace &#39;{}&#39; already exists".format(name)
+        if response_code == 201:
+            LOGGER.info("Geoserver workspace [%s] was successfully created.", name)
+        elif response_code == 401 and string_to_find in response.text:
+            # This is done because Geoserver's reply/error code is misleading in this case and
+            # returns HTML content.
+            LOGGER.error("The following Geoserver workspace already exists: [%s]", name)
+        elif response_code == 401:
+            LOGGER.error("The request has not been applied because it lacks valid authentication credentials.")
+        elif response_code == 500:
+            LOGGER.error(response.text)
+        else:
+            LOGGER.error("There was an error creating the workspace in Geoserver : %s", name)
+
+    def remove_workspace(self, name):
+        # type (Geoserver, str) -> None
+        """
+        Removes a workspace from geoserver. Will also remove all datastores associated with
+        the workspace.
+
+        @param name: Workspace name
+        """
+        LOGGER.info("Attempting to remove workspace in geoserver")
+        response = self._remove_workspace_request(name)
+        response_code = response.status_code
+        if response_code == 200:
+            LOGGER.info("Geoserver workspace [%s] was successfully removed.", name)
+        elif response_code == 403:
+            LOGGER.error(
+                "Geoserver workspace [%s] is not empty. Make sure `recurse` is set to `true` to delete workspace")
+        elif response_code == 404:
+            LOGGER.error("Geoserver workspace [%s] was not found.", name)
+
+    def create_datastore(self, workspace_name):
+        # type (Geoserver, str) -> None
         """
         Create a new Geoserver workspace.
 
         @param self: Geoserver instance
-        @param workspace_id: Workspace id where the datastore must be created
-        @param name: Datastore name
-        @return: Datastore id
+        @param workspace_name: Workspace name where the datastore must be created
         """
         LOGGER.info("Creating datastore in geoserver")
-        # TODO
-        return 1
+
+        datastore_name = self.get_datastore_name(workspace_name)
+        creation_response = self._create_datastore_request(workspace_name=workspace_name, datastore_name=datastore_name)
+        response_code = creation_response.status_code
+        if response_code == 201:
+            LOGGER.info("Datastore [%s] has been successfully created.", datastore_name)
+        elif response_code == 401:
+            LOGGER.error("The request has not been applied because it lacks valid authentication credentials.")
```

Comment:
Does it always return 401 even when auth are provided but are incorrect? Not 403 instead?

Reply:
Yes, Geoserver always returns 401 for either missing or incorrect credentials

Reply:
Can you add a note about this because this is not the same behaviour across all API.

Reply:
Read up a bit on this, I guess it depends what you mean by incorrect. From my understanding, 403 is for when a user credentials are valid, but does not have permission to access a specific ressource (permanently forbidden), whereas 401 is for a lack or bad credentials (say, typo in the password).

In our use case, it should always be the Geoserver admin user, who has all the permissions necessary, so the only 403 we should get would be related to specific routes and their restrictions, like here : https://github.com/Ouranosinc/cowbird/blob/geoserver-setup/cowbird/services/impl/geoserver.py#L56

I did add a generic `else` to `geoserver_response_handling` decorator so we'll still catch the error codes with the operations that lead up to them for all the cases that fall outside of what I tested.

Reply:
Yes, your understanding of 401/403 is the correct one. 
I know some API frameworks incorrectly replace 403 by 401, or vice-versa in some cases. 
Naming conventions of "Unauthorized" vs "Forbidden" are also often swapped around erroneously.
This is why I was wondering if 403 was not needed also in this case.